### DISCUSSION TITLE
Rust bindings

### DISCUSSION THREAD

**Discussion by prideout** [elalish/manifold#1046] [General]
Word on the street is that someone is writing Rust bindings? Curious how far along that is, and if I can help in anyway? At Arcol we want to use manifold in a Rust-on-WASM scenario.

@ncthbrt @NickUfer

**Comment by NickUfer** [comment]
I'm working on it, but it is currently paused until the API changes for V3.0 are somewhat final.

This is the repo responsible for building manifold and providing raw ffi bindings: https://github.com/NickUfer/manifold-sys-rs
This is the repo of the library providing the safe API: https://github.com/NickUfer/manifold3d-rs

So far I did not try to build the project for wasm yet, I think there will be some hurdles too. There is also an open problem on the sys-rs crate regarding crate size I put on the side for now: https://github.com/NickUfer/manifold-sys-rs/issues/1

**Reply by prideout** [reply]
Interesting, thanks!  We certainly would not want to include `assimp` in our WASM bundle, maybe that's something I can help with.

**Reply by NickUfer** [reply]
Yeah that is something we can hide behind a feature flag, which in turn controls the feature flag `MANIFOLD_EXPORT`. I just added it for now to include all features and make this optional later.

**Reply by pca006132** [reply]
@NickUfer I think it may be much simpler to use the [cmake crate](https://docs.rs/cmake/latest/cmake/) to build manifold, and supply the dependencies by specifying `MANIFOLD_USE_BUILTIN_TBB` and `FETCHCONTENT_SOURCE_DIR_TBB` for example. Our cmake already handles the cross-platform shenanigans, so you don't have to do it again :)

Regarding crate size, apparently other -sys crates cheat by downloading it elsewhere instead of using submodule: https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/build/find_vendored.rs

And for `assimp`, I think if you depend on the system installed version (I don't think you are building assimp according to your code), you should just somehow find the package by either using cmake environment variables or pkg-config instead of pulling in the assimp git just for a single header file. Again, the package search is already handled by cmake...

**Comment by NickUfer** [comment]
Yeah it would probably be easier to use cmake, but alot of rust devy complain about build errors when using cmake, especially on Windows. The last thing rust devs want to deal with is build errors of some C/C++ dependency. Thats why I am trying to use cc, as it is just working on linux and windows with minmal effort. 

About crate size: I actually managed to fix it today by explicitly listing all files/folders which should be included in the upload. This reduced the overall size to ~650KB compressed, which is acceptable.

About assimp and the other libraries: I used static linking for now because I knew this will work and I figure this out on the way. But of course dynamic linking and just linking against the installed version in the environment is the next goal.

**Comment by NickUfer** [comment]
An update on WASM support: Right now, I think it is very hard to support compilation for WASM in any way. There are different compilation targets available, like `wasm32-unknown-unknown` (most supported), `wasm32-unknown-emscripten` (on life support), etc. 
Compilation for the emscripten target works, but the rest of the tooling like `wasm-pack` and `wasm_bindgen` only supports `wasm32-unknown-unknown`. `wasm32-unknown-unknown` does not support any C/C++ interop in any way.
There is experimental support for the C ABI, and the library only uses the C bindings, but I don't know if this is enough.

One possibility would be to write all the JS/TS binding code manually, but I wan't to avoid that.

**Comment by pca006132** [comment]
WASM may not actually be that hard. I managed to compile to wasm using just clang and target `wasm32-unknown-unknown`, with some modification and some other libraries, and get to some acceptable number of missing symbols...

Basically:

```sh
MANIFOLD="manifold"
LIBCXX="llvm-project/libcxx" # we need the source because we need to compile several source files from here
WASMLIBC="mono-wasm-libc"
CLIPPER2="Clipper2"

INCLUDES="-isystem $LIBCXX/include -isystem $WASMLIBC/include -isystem $WASMLIBC/arch/wasm32 -I./ -I$CLIPPER2/CPP/Clipper2Lib/include -I$MANIFOLD/include -I$MANIFOLD/bindings/c/include "
WASMFLAGS="--target=wasm32-unknown-unknown -fno-exceptions -fno-rtti -nostdlib"

for f in $MANIFOLD/src/*.cpp $MANIFOLD/src/cross_section/*.cpp $MANIFOLD/bindings/c/*.cpp $CLIPPER2/CPP/Clipper2Lib/src/*.cpp $LIBCXX/src/{bind,hash,memory,new_helpers}.cpp; do
  if [ "$(basename $f)" = "meshIOc.cpp" ]; then
    continue
  fi
  echo compiling $f
  clang -DCLIPPER2_MAX_DECIMAL_PRECISION=8 -DMANIFOLD_EXPORTS -DMANIFOLD_CROSS_SECTION -DMANIFOLD_PAR=-1 $INCLUDES -DNDEBUG -std=c++17 -Wall -Wno-unused -Werror $WASMFLAGS -o $(basename ${f/.cpp/.o}) -c $f
done

wasm-ld --no-entry --export-all --warn-unresolved-symbols *.o -o libmanifold.wasm 2>&1 | grep -o "undefined symbol: .*" | sort --unique
```

Requires the following modifications for `mono-wasm-libc`, `Clipper2` as well as manifold (currently in branch https://github.com/elalish/manifold/tree/wasm32-unknown-unknown):

<details>

mono-wasm-libc: clang wants `size_t` to be `unsigned long`.

```diff
diff --git a/arch/wasm32/bits/alltypes.h b/arch/wasm32/bits/alltypes.h
index 06a3049e..a992894d 100644
--- a/arch/wasm32/bits/alltypes.h
+++ b/arch/wasm32/bits/alltypes.h
@@ -1,6 +1,6 @@
-#define _Addr int
+#define _Addr long
 #define _Int64 long long
-#define _Reg int
+#define _Reg long
 
 #if defined(__NEED_va_list) && !defined(__DEFINED_va_list)
 typedef __builtin_va_list va_list;
@@ -79,6 +79,11 @@ typedef unsigned _Addr size_t;
 #define __DEFINED_size_t
 #endif
 
+#if defined(__NEED_max_align_t) && !defined(__DEFINED_max_align_t)
+typedef struct { long long __ll; long double __ld; } max_align_t;
+#define __DEFINED_max_align_t
+#endif
+
 #if defined(__NEED_uintptr_t) && !defined(__DEFINED_uintptr_t)
 typedef unsigned _Addr uintptr_t;
 #define __DEFINED_uintptr_t
```

Clipper2: Remove iostream

```diff
diff --git a/CPP/Clipper2Lib/include/clipper2/clipper.core.h b/CPP/Clipper2Lib/include/clipper2/clipper.core.h
index 624de90..4ee072d 100644
--- a/CPP/Clipper2Lib/include/clipper2/clipper.core.h
+++ b/CPP/Clipper2Lib/include/clipper2/clipper.core.h
@@ -15,7 +15,6 @@
 #include <cmath>
 #include <vector>
 #include <string>
-#include <iostream>
 #include <algorithm>
 #include <climits>
 #include <numeric>
@@ -169,11 +168,11 @@ namespace Clipper2Lib
 
     void SetZ(const z_type z_value) { z = z_value; }
 
-    friend std::ostream& operator<<(std::ostream& os, const Point& point)
-    {
-      os << point.x << "," << point.y << "," << point.z;
-      return os;
-    }
+    // friend std::ostream& operator<<(std::ostream& os, const Point& point)
+    // {
+    //   os << point.x << "," << point.y << "," << point.z;
+    //   return os;
+    // }
 
 #else
 
@@ -206,11 +205,11 @@ namespace Clipper2Lib
       return Point(x * scale, y * scale);
     }
 
-    friend std::ostream& operator<<(std::ostream& os, const Point& point)
-    {
-      os << point.x << "," << point.y;
-      return os;
-    }
+    // friend std::ostream& operator<<(std::ostream& os, const Point& point)
+    // {
+    //   os << point.x << "," << point.y;
+    //   return os;
+    // }
 #endif
 
     friend bool operator==(const Point& a, const Point& b)
@@ -385,10 +384,10 @@ namespace Clipper2Lib
       return result;
     }
 
-    friend std::ostream& operator<<(std::ostream& os, const Rect<T>& rect) {
-      os << "(" << rect.left << "," << rect.top << "," << rect.right << "," << rect.bottom << ") ";
-      return os;
-    }
+    // friend std::ostream& operator<<(std::ostream& os, const Rect<T>& rect) {
+    //   os << "(" << rect.left << "," << rect.top << "," << rect.right << "," << rect.bottom << ") ";
+    //   return os;
+    // }
   };
 
   template <typename T1, typename T2>
@@ -487,26 +486,26 @@ namespace Clipper2Lib
     return Rect<T>(xmin, ymin, xmax, ymax);
   }
 
-  template <typename T>
-  std::ostream& operator << (std::ostream& outstream, const Path<T>& path)
-  {
-    if (!path.empty())
-    {
-      auto pt = path.cbegin(), last = path.cend() - 1;
-      while (pt != last)
-        outstream << *pt++ << ", ";
-      outstream << *last << std::endl;
-    }
-    return outstream;
-  }
-
-  template <typename T>
-  std::ostream& operator << (std::ostream& outstream, const Paths<T>& paths)
-  {
-    for (auto p : paths)
-      outstream << p;
-    return outstream;
-  }
+  // template <typename T>
+  // std::ostream& operator << (std::ostream& outstream, const Path<T>& path)
+  // {
+  //   if (!path.empty())
+  //   {
+  //     auto pt = path.cbegin(), last = path.cend() - 1;
+  //     while (pt != last)
+  //       outstream << *pt++ << ", ";
+  //     outstream << *last << std::endl;
+  //   }
+  //   return outstream;
+  // }
+  //
+  // template <typename T>
+  // std::ostream& operator << (std::ostream& outstream, const Paths<T>& paths)
+  // {
+  //   for (auto p : paths)
+  //     outstream << p;
+  //   return outstream;
+  // }
 
 
   template <typename T1, typename T2>
diff --git a/CPP/Clipper2Lib/include/clipper2/clipper.engine.h b/CPP/Clipper2Lib/include/clipper2/clipper.engine.h
index 559f247..3e73816 100644
--- a/CPP/Clipper2Lib/include/clipper2/clipper.engine.h
+++ b/CPP/Clipper2Lib/include/clipper2/clipper.engine.h
@@ -12,7 +12,7 @@
 
 #include <cstdlib>
 #include <stdint.h> //#541
-#include <iostream>
+// #include <iostream>
 #include <queue>
 #include <vector>
 #include <functional>
diff --git a/CPP/Clipper2Lib/include/clipper2/clipper.h b/CPP/Clipper2Lib/include/clipper2/clipper.h
index a2fe5c3..65a059d 100644
--- a/CPP/Clipper2Lib/include/clipper2/clipper.h
+++ b/CPP/Clipper2Lib/include/clipper2/clipper.h
@@ -311,35 +311,35 @@ namespace Clipper2Lib {
       return true;
     }
 
-    static void OutlinePolyPath(std::ostream& os,
-      size_t idx, bool isHole, size_t count, const std::string& preamble)
-    {
-      std::string plural = (count == 1) ? "." : "s.";
-      if (isHole)
-        os << preamble << "+- Hole (" << idx << ") contains " << count <<
-        " nested polygon" << plural << std::endl;
-      else
-        os << preamble << "+- Polygon (" << idx << ") contains " << count <<
-          " hole" << plural << std::endl;
-    }
-
-    static void OutlinePolyPath64(std::ostream& os, const PolyPath64& pp,
-      size_t idx, std::string preamble)
-    {
-      OutlinePolyPath(os, idx, pp.IsHole(), pp.Count(), preamble);
-      for (size_t i = 0; i < pp.Count(); ++i)
-        if (pp.Child(i)->Count())
-          details::OutlinePolyPath64(os, *pp.Child(i), i, preamble + "  ");
-    }
-
-    static void OutlinePolyPathD(std::ostream& os, const PolyPathD& pp,
-      size_t idx, std::string preamble)
-    {
-      OutlinePolyPath(os, idx, pp.IsHole(), pp.Count(), preamble);
-      for (size_t i = 0; i < pp.Count(); ++i)
-        if (pp.Child(i)->Count())
-          details::OutlinePolyPathD(os, *pp.Child(i), i, preamble + "  ");
-    }
+    // static void OutlinePolyPath(std::ostream& os,
+    //   size_t idx, bool isHole, size_t count, const std::string& preamble)
+    // {
+    //   std::string plural = (count == 1) ? "." : "s.";
+    //   if (isHole)
+    //     os << preamble << "+- Hole (" << idx << ") contains " << count <<
+    //     " nested polygon" << plural << std::endl;
+    //   else
+    //     os << preamble << "+- Polygon (" << idx << ") contains " << count <<
+    //       " hole" << plural << std::endl;
+    // }
+    //
+    // static void OutlinePolyPath64(std::ostream& os, const PolyPath64& pp,
+    //   size_t idx, std::string preamble)
+    // {
+    //   OutlinePolyPath(os, idx, pp.IsHole(), pp.Count(), preamble);
+    //   for (size_t i = 0; i < pp.Count(); ++i)
+    //     if (pp.Child(i)->Count())
+    //       details::OutlinePolyPath64(os, *pp.Child(i), i, preamble + "  ");
+    // }
+    //
+    // static void OutlinePolyPathD(std::ostream& os, const PolyPathD& pp,
+    //   size_t idx, std::string preamble)
+    // {
+    //   OutlinePolyPath(os, idx, pp.IsHole(), pp.Count(), preamble);
+    //   for (size_t i = 0; i < pp.Count(); ++i)
+    //     if (pp.Child(i)->Count())
+    //       details::OutlinePolyPathD(os, *pp.Child(i), i, preamble + "  ");
+    // }
 
     template<typename T, typename U>
     inline constexpr void MakePathGeneric(const T an_array,
@@ -356,28 +356,28 @@ namespace Clipper2Lib {
 
   } // end details namespace
 
-  inline std::ostream& operator<< (std::ostream& os, const PolyTree64& pp)
-  {
-    std::string plural = (pp.Count() == 1) ? " polygon." : " polygons.";
-    os << std::endl << "Polytree with " << pp.Count() << plural << std::endl;
-      for (size_t i = 0; i < pp.Count(); ++i)
-        if (pp.Child(i)->Count())
-          details::OutlinePolyPath64(os, *pp.Child(i), i, "  ");
-    os << std::endl << std::endl;
-    return os;
-  }
-
-  inline std::ostream& operator<< (std::ostream& os, const PolyTreeD& pp)
-  {
-    std::string plural = (pp.Count() == 1) ? " polygon." : " polygons.";
-    os << std::endl << "Polytree with " << pp.Count() << plural << std::endl;
-    for (size_t i = 0; i < pp.Count(); ++i)
-      if (pp.Child(i)->Count())
-        details::OutlinePolyPathD(os, *pp.Child(i), i, "  ");
-    os << std::endl << std::endl;
-    if (!pp.Level()) os << std::endl;
-    return os;
-  }
+  // inline std::ostream& operator<< (std::ostream& os, const PolyTree64& pp)
+  // {
+  //   std::string plural = (pp.Count() == 1) ? " polygon." : " polygons.";
+  //   os << std::endl << "Polytree with " << pp.Count() << plural << std::endl;
+  //     for (size_t i = 0; i < pp.Count(); ++i)
+  //       if (pp.Child(i)->Count())
+  //         details::OutlinePolyPath64(os, *pp.Child(i), i, "  ");
+  //   os << std::endl << std::endl;
+  //   return os;
+  // }
+  //
+  // inline std::ostream& operator<< (std::ostream& os, const PolyTreeD& pp)
+  // {
+  //   std::string plural = (pp.Count() == 1) ? " polygon." : " polygons.";
+  //   os << std::endl << "Polytree with " << pp.Count() << plural << std::endl;
+  //   for (size_t i = 0; i < pp.Count(); ++i)
+  //     if (pp.Child(i)->Count())
+  //       details::OutlinePolyPathD(os, *pp.Child(i), i, "  ");
+  //   os << std::endl << std::endl;
+  //   if (!pp.Level()) os << std::endl;
+  //   return os;
+  // }
 
   inline Paths64 PolyTreeToPaths64(const PolyTree64& polytree)
   {
```

As well as the following two files for libcxx:

```c++
// __config_site
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#ifndef _LIBCPP___CONFIG_SITE
#define _LIBCPP___CONFIG_SITE

#define _LIBCPP_ABI_VERSION 1
#define _LIBCPP_ABI_NAMESPACE __1
#define _LIBCPP_HAS_NO_THREADS
#define _LIBCPP_HAS_NO_MONOTONIC_CLOCK
/* #define _LIBCPP_HAS_MUSL_LIBC */
/* #undef _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS */
#define _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS
/* #undef _LIBCPP_TYPEINFO_COMPARISON_IMPLEMENTATION */
#define _LIBCPP_HAS_NO_FILESYSTEM
#define _LIBCPP_HAS_NO_RANDOM_DEVICE
#define _LIBCPP_HAS_NO_LOCALIZATION
#define _LIBCPP_HAS_NO_WIDE_CHARACTERS
#define _LIBCPP_HAS_NO_STD_MODULES
#define _LIBCPP_HAS_NO_TIME_ZONE_DATABASE
/* #undef _LIBCPP_INSTRUMENTED_WITH_ASAN */

// PSTL backends
#define _LIBCPP_PSTL_CPU_BACKEND_SERIAL
/* #undef _LIBCPP_PSTL_CPU_BACKEND_THREAD */
/* #undef _LIBCPP_PSTL_CPU_BACKEND_LIBDISPATCH */

// Hardening.
#define _LIBCPP_HARDENING_MODE_DEFAULT 2

// __USE_MINGW_ANSI_STDIO gets redefined on MinGW
#ifdef __clang__
#  pragma clang diagnostic push
#  pragma clang diagnostic ignored "-Wmacro-redefined"
#endif

#ifdef __clang__
#  pragma clang diagnostic pop
#endif

#endif // _LIBCPP___CONFIG_SITE
```

```c++
// __assertion_handler
// -*- C++ -*-
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#ifndef _LIBCPP___ASSERTION_HANDLER
#define _LIBCPP___ASSERTION_HANDLER

#include <__config>
#include <__verbose_abort>

#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
#  pragma GCC system_header
#endif

#if _LIBCPP_HARDENING_MODE == _LIBCPP_HARDENING_MODE_DEBUG

#  define _LIBCPP_ASSERTION_HANDLER(message) _LIBCPP_VERBOSE_ABORT("%s", message)

#else

// TODO(hardening): use `__builtin_verbose_trap(message)` once that becomes available.
#  define _LIBCPP_ASSERTION_HANDLER(message) ((void)message, __builtin_trap())

#endif // _LIBCPP_HARDENING_MODE == _LIBCPP_HARDENING_MODE_DEBUG

#endif // _LIBCPP___ASSERTION_HANDLER

```

</details>

Anyway, at the end of the day, these are the remaining undefined symbols:
```
undefined symbol: acos
undefined symbol: asin
undefined symbol: atan2
undefined symbol: cos
undefined symbol: __cxa_atexit
undefined symbol: __cxa_pure_virtual
undefined symbol: fma
undefined symbol: fmax
undefined symbol: fmin
undefined symbol: free
undefined symbol: hypot
undefined symbol: ilogb
undefined symbol: log10
undefined symbol: log2
undefined symbol: malloc
undefined symbol: memcpy
undefined symbol: memmove
undefined symbol: memset
undefined symbol: operator delete(void*)
undefined symbol: operator delete[](void*)
undefined symbol: operator delete(void*, std::align_val_t)
undefined symbol: operator new(unsigned long)
undefined symbol: operator new[](unsigned long)
undefined symbol: operator new(unsigned long, std::align_val_t)
undefined symbol: operator new(unsigned long, std::nothrow_t const&)
undefined symbol: pow
undefined symbol: remquo
undefined symbol: round
undefined symbol: sin
undefined symbol: std::__1::__libcpp_verbose_abort(char const*, ...)
undefined symbol: std::exception::~exception()
```

There are basically 3 categories:
1. Math functions provided by libm, rust already has a libm crate for exactly the same purposes.
2. Memory related functions, should be simple to write a shim.
3. Functions that are only called when panic/cleanup, which we probably don't have to care and can write a simple shim.

I have no idea if this will work properly even when linked, especially when we have static variables that we should init. But one can probably write some simple c++ code to try...

**Reply by NickUfer** [reply]
Hmm, interesting!
When I try to compile for wasm32-unkown-unknown I get this error:

```
  cargo:warning=vendor/Clipper2/CPP/Clipper2Lib/src/clipper.engine.cpp:10:10: fatal error: 'cstdlib' file not found
  cargo:warning=   10 | #include <cstdlib>
  cargo:warning=      |          ^~~~~~~~~
  cargo:warning=1 error generated.

  --- stderr


  error occurred: Command "clang++" "-O0" "-ffunction-sections" "-fdata-sections" "-fno-exceptions" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-I" "vendor/Clipper2/CPP/Clipper2Lib/include" "-I" "vendor/Clipper2/CPP/Utils" "-Wall" "-Wextra" "-std=c++17" "-o" "/home/nick/RustroverProjects/manifold3d-sys/target/wasm32-unknown-unknown/debug/build/manifold3d-sys-9138bb83238c6392/out/8e38fc333959fbb2-clipper.engine.o" "-c" "vendor/Clipper2/CPP/Clipper2Lib/src/clipper.engine.cpp" with args clang++ did not execute successfully (status code exit status: 1).
```

I will quickly try your branch

**Reply by pca006132** [reply]
You need a lot more system include flags, basically building a partial sysroot for wasm-unknown-unknown.

**Reply by pca006132** [reply]
No, my branch will not help much, the branch just reduces the number of libcxx files we have to compile to get every symbol satisfied.

**Reply by NickUfer** [reply]
Okay, I guess I will focus for now on implementing the rust API and hope they add wasm32-unknown-emscripten support for the tools I need to generate the JS glue code. There is already an effort to do this going on here: https://github.com/rustwasm/wasm-bindgen/pull/4014

**Comment by elalish** [comment]
@WilstonOreo [mentioned](https://github.com/elalish/manifold/issues/1117#issuecomment-2557604233) they also have some WIP Rust bindings. Perhaps you all can find a way to merge your efforts and build a complete set of bindings.

**Comment by NickUfer** [comment]
Funny, two guys from Berlin working on the same thing.
Doesn't look like it is combinable, he is writing a wrapper around C++ , I am focusing around the stable C ABI and leverage types to prevent as much as possible input of invalid data.

They also seem to build it with a bigger intend, to use it in MicroCAD. Such things always make me wonder why should I invest any more time in my small side project when something like this will be much more supported and probably used because it is used in some other bigger project. I guess I'll put my lib on the back burner for now and focus on other stuff.

Am Samstag, Dezember 21, 2024 18:41 CET, schrieb Emmett Lalish ***@***.***>:

 
 
@WilstonOreo mentioned they also have some WIP Rust bindings. Perhaps you all can find a way to merge your efforts and build a complete set of bindings.
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>
[ { ***@***.***": "http://schema.org", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "https://github.com/elalish/manifold/discussions/1046#discussioncomment-11638717", "url": "https://github.com/elalish/manifold/discussions/1046#discussioncomment-11638717", "name": "View Discussion" }, "description": "View this Discussion on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

**Reply by WilstonOreo** [reply]
Yes, we use manifold-rs for [µcad](https://github.com/Rustfahrtagentur/microcad/), which is project from our Rust learning group at Raumfahrtagentur.org. Not sure how far we get with it, but we are very ambitious :)

**Comment by prideout** [comment]
Quick update on my end: our web app is currently using manifold by calling into JS from our Rust WASM module, and our JS layer in turn calls into Manifold's published WASM bundle.  It'll be nice to remove the indirection.

**Reply by WilstonOreo** [reply]
Nice, I haven't tested my library on WASM yet, maybe you can give it a try :)

**Comment by zmerlynn** [comment]
Thanks @pca006132 — picking up this thread: your recipe became [wasm-cxx-shim](https://github.com/zmerlynn/wasm-cxx-shim), a minimal `libc`/`libm`/`libcxx` for `wasm32-unknown-unknown`. The 31-symbol shopping list and `__config_site` override both still apply.

End-to-end status: manifold v3.4.1's library + **71/71 of its own [tests](https://github.com/zmerlynn/wasm-cxx-shim/blob/main/test/manifold-tests/README.md)** (`boolean_test`, `sdf_test`, `cross_section_test`) run on the shim under Node, **zero unexpected wasm imports**. Rust binding side: [`zmerlynn/manifold-csg`](https://github.com/zmerlynn/manifold-csg) builds + runs on `wasm32-unknown-unknown` against the shim — shipped experimentally in [manifold-csg 0.1.7](https://crates.io/crates/manifold-csg/0.1.7) — which should make manifold consumable from `wasm-bindgen`-based stacks (Bevy, Leptos, Yew, …) directly. Demo: <https://zmerlynn.github.io/manifold-csg/>.

Three small carry-patches make the integration work — would PRs guarding them with build-time macros be welcome upstream? Each leaves upstream behavior unchanged when the macro isn't set:

- [`0002-manifold-ifdef-iostream.patch`](https://github.com/zmerlynn/wasm-cxx-shim/blob/main/cmake/manifold-patches/0002-manifold-ifdef-iostream.patch) — `MANIFOLD_NO_IOSTREAM` for OBJ I/O in `src/impl.cpp` + the `manifold_*_obj` bindings.
- [`0003-manifold-test-main-ifdef-filesystem.patch`](https://github.com/zmerlynn/wasm-cxx-shim/blob/main/cmake/manifold-patches/0003-manifold-test-main-ifdef-filesystem.patch) — `MANIFOLD_NO_FILESYSTEM` for `test/test_main.cpp`'s `<filesystem>` / `<fstream>` / `main()`, with stub-alternate fixture helpers so consumers keep compiling the test sources directly.
- [`0001-clipper2-strip-iostream.patch`](https://github.com/zmerlynn/wasm-cxx-shim/blob/main/cmake/manifold-patches/0001-clipper2-strip-iostream.patch) — `CLIPPER2_NO_IOSTREAM` for Clipper2's `operator<<` / outline-poly-path helpers. Since manifold pins Clipper2 via FetchContent at a specific SHA, this presumably routes through AngusJohnson/Clipper2 first then we'd bump the pin here once it lands — worth checking with you in case you'd prefer different routing.

If anyone wants to try it today: `find_package(wasm-cxx-shim)` exposes a [`wasm_cxx_shim_add_manifold()`](https://github.com/zmerlynn/wasm-cxx-shim/blob/main/cmake/WasmCxxShimManifold.cmake) helper that bundles the pinned manifold + Clipper2 versions, applies the patches, and flips manifold's CMake options.

**Reply by pca006132** [reply]
I am open to these patches. Ideally we want clipper2 to patch upstream. Before that, I prefer to use cmake/CI to apply a patch file to clipper2 after pulling in the dependency.

**Reply by zmerlynn** [reply]
Implementation up at #1690: opt-in CMake option + tracked Clipper2 carry-patch applied via `FetchContent`'s `PATCH_COMMAND` (mirrors AngusJohnson/Clipper2#1094, drops once that lands).

**Comment by elalish** [comment]
Checkout [manifold-rust](https://github.com/larsbrubaker/manifold-rust) by @larsbrubaker, discussed over in #1774. I'm super impressed to see a complete Manifold port to native Rust that passes every test identically and pretty much matches performance. Fantastic work, and I'd love to hear how it works for the users here that have been hoping for such a thing to exist.

**Reply by WilstonOreo** [reply]
I integrated manifold-rust last week into [microcad](https://codeberg.org/microcad/microcad/) and it works like a charm!
I have archived my [manifold-rs](https://github.com/WilstonOreo/manifold-rs/) bindings.

**Comment by larsbrubaker** [comment]
Hey @WilstonOreo,

It's great to hear that you are having success with the rest version. Have you seen that we also have support for non-manifold meshes in there? It is a direct port of the research paper that was used in libigl, the basis of the Blender version of robust non-manifold Boolean intersections.

**Reply by WilstonOreo** [reply]
No, I haven't tested the robust kernel yet, but will in the upcoming weeks. We had indeed several issues with non-manifold meshes, so I am excited to verify if manifold-rust can handle that.
